### PR TITLE
7357 filestore keyword v2.0

### PR DESCRIPTION
**Pull request by regit**
Update of #12073

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to tickets:
- https://redmine.openinfosecfoundation.org/issues/7356
- https://redmine.openinfosecfoundation.org/issues/7357

Describe changes:
- Rebase to master
- Address comments from #12073

SV_REPO=https://github.com/OISF/suricata-verify/pull/2111

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23597

File: src/detect-engine-file.c [hunk 1]
```diff
@@ -194,7 +194,15 @@ uint8_t DetectFileInspectGeneric(DetectEngineCtx *de_ctx, DetectEngineThreadCtx
     if (ffc == NULL) {
         SCReturnInt(DETECT_ENGINE_INSPECT_SIG_CANT_MATCH_FILES);
     } else if (ffc->head == NULL) {
-        SCReturnInt(DETECT_ENGINE_INSPECT_SIG_NO_MATCH);
+        if (s->flags & SIG_FLAG_FILESTORE) {
```
Comment (catenacyber) [inline src/detect-engine-file.c:197]: Will this ignore the `file.data` part in 
`alert http any any -> any any (http.uri; content: "toto"; filestore; file.data; content: "tata";)` and match for GET /toto when it should match only for `POST /toto ... tata`

**Comment by catenacyber** [comment]
I think https://github.com/OISF/suricata/pull/11900 may help here : you want to match on uri and store the downloaded file, right ?

**Comment by victorjulien** [comment]
"The filestore keyword had an influence on the signature matching
when it should not."

This too vague. Please describe what the influence is in the commit, why it is wrong, and how the commit addresses this.

**Comment by regit** [comment]
> I think #11900 may help here : you want to match on uri and store the downloaded file, right ?

That is a bit the same problem as rule will want to trigger something in the other direction but it is different. In the case of this PR, I'm just implementing options that were designed from the beginning of filestore keyword (but not documented I think). For example, it allows to write a signature where we say "extract all the files that are an answer to a request on this url":
 
```
alert http any any -> any any (msg:"dl url"; http.uri:"/dl"; filestore:to_client,tx;)
```

File: src/detect-engine-file.c (hunk 1 above)
Comment (regit) [reply src/detect-engine-file.c:197]: Really interesting point to check. I'm going to cook suricata-verify test for that and correct current code if necessary.

Comment (regit) [reply src/detect-engine-file.c:197]: test added and it was working fine.

**Pull request closed without merging**